src/Gui/MainWindow.cpp

```diff
@@ -2222,6 +2230,128 @@ void MainWindow::showStatus(int type, const QString& message)
     statusBar()->showMessage(msg.simplified(), timeout);
 }
 
+void MainWindow::showHint(InputHint hint)
+{
+    showHints({ hint });
+}
+
+void MainWindow::showHints(const std::list<InputHint>& hints)
```

Comment:
Why not put the image generation into `InputHint.h`? That would group related logic, and keep `MainWindow.cpp` clean/minimal/less of a mess.

Then some of the lambdas could also be made static functions... to me, it would improve readability/overview as well

Reply:
Right, I wanted that struct to be strictly data but the logic indeed is quite large right now. I'll think on how it could be moved outside of the main window as you are totally right that it is not the right place to do it.

Maybe custom widget inheriting from QLabel with that logic would a good choice - it would then encapsulate that logic and keep everything together.